core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java

```diff
@@ -0,0 +1,271 @@
+     * There's no point to use a LongAdder here, because reading the counters would hit worse than sharing a single one
+     * which can leverage fast x86 fetch-and-add instructions.
+     */
+    private static final int INTS_PER_CACHE_LINE = 64 / Integer.BYTES;
+    private static final int USED_OFFSET = INTS_PER_CACHE_LINE;
+    private static final int MIN_USED_OFFSET = INTS_PER_CACHE_LINE * 2;
+    private static final int MAX_USED_OFFSET = INTS_PER_CACHE_LINE * 3;
+    private static final int COUNTERS_INTS = INTS_PER_CACHE_LINE * 4;
+    private final AtomicIntegerArray counters;
+
+
+    public AffinityAwareSessionPool(EventExecutor[] eventExecutors, Supplier<Session> sessionSupplier) {
+        this.sessionSupplier = sessionSupplier;
+        this.agentThreadIdPerExecutor = new IdentityHashMap<>(eventExecutors.length);
+        this.localQueues = new Queue[eventExecutors.length];
+        for (int agentThreadId = 0; agentThreadId < eventExecutors.length; agentThreadId++) {
+            var boxedAgentThreadId = agentThreadId;
+            agentThreadIdPerExecutor.put(eventExecutors[agentThreadId], boxedAgentThreadId);
+        }
+        this.localAgentThreadId = new FastThreadLocal<>() {
+            @Override
+            protected Integer initialValue() {
+                var eventExecutor = ThreadExecutorMap.currentExecutor();
+                if (eventExecutor == null) {
+                    return null;
+                }
+                var agentThreadId = agentThreadIdPerExecutor.get(eventExecutor);
+                if (agentThreadId == null) {
+                    return null;
+                }
+                return agentThreadId;
+            }
+        };
+        this.counters = new AtomicIntegerArray(COUNTERS_INTS);
+    }
+
+    private int nextWorkStealIndex() {
+        return ThreadLocalRandom.current().nextInt(0, localQueues.length);
+    }
+
+    private int nextWorkStealIndex(int excludeIndex) {
+        var localQueuesCount = localQueues.length;
+        var workStealIndex = ThreadLocalRandom.current().nextInt(0, localQueuesCount);
+        if (workStealIndex == excludeIndex) {
+            workStealIndex++;
+            if (workStealIndex == localQueuesCount) {
+                workStealIndex = 0;
+            }
+        }
+        return workStealIndex;
+    }
+
+
+    @Override
+    public Session acquire() {
+        var boxedAgentThreadId = localAgentThreadId.get();
+        if (boxedAgentThreadId == null) {
+            return acquireFromLocalQueues();
+        }
+        var agentThreadId = boxedAgentThreadId.intValue();
+        var localQueues = this.localQueues;
+        var session = localQueues[agentThreadId].poll();
+        if (session != null) {
+            incrementUsed();
+            return session;
+        }
+        if (localQueues.length == 1) {
+            return null;
+        }
+        return acquireFromOtherLocalQueues(localQueues, agentThreadId);
```

Comment:
Couple of questions here:
- Why do we need explicit unboxing for `boxedAgentThreadId`? isn't it already automatically managed?
- Why do we need to save `this.localQueues` to a local variable? is there any particular reason or benefit?

If there is no particular reason fo rboth doubts I think we can simplify a little bit:
```suggestion
        var boxedAgentThreadId = localAgentThreadId.get();
        if (boxedAgentThreadId == null) {
            return acquireFromLocalQueues();
        }
        var session = localQueues[boxedAgentThreadId].poll();
        if (session != null) {
            incrementUsed();
            return session;
        }
        if (this.localQueues.length == 1) {
            return null;
        }
        return acquireFromOtherLocalQueues(this.localQueues, boxedAgentThreadId);
```

Reply:
The main reason was to avoid twice `Integer::intValue`: given that there are atomic operations in-between the 2 unboxing operations, the JIT can decide to actual perform it twice, without trusting that Integer's value hasn't changed :/